src/components/scripts/oauth-connector.ts

```diff
@@ -0,0 +1,241 @@
+    this.#promiseResolver = resolve;
+    this.#promiseRejecter = reject;
+
+    this.#socketServer.opened.then((server) => {
+      this.#redirectUri = `http://localhost:${server.localPort}`;
+      this.#runHttpServer();
+      this.#authenticate(this.#authenticationEndpoint);
+    });
+
+    return promise;
+  }
+
+  get authentication_token(): string | undefined {
+    return this.#authenticationCode;
+  }
+
+  get access_token(): string | undefined {
+    return this.#accessCode;
+  }
+
+  get refresh_token(): string | undefined {
+    return this.#refreshCode;
+  }
+
+  #authenticate(endpoint: string): void {
+    // Create <form> element to submit parameters to OAuth 2.0 endpoint.
+    const form = document.createElement('form');
+    form.setAttribute('method', 'GET'); // Send as a GET request.
+    form.setAttribute('action', endpoint);
+    form.setAttribute('target', '_blank');
+
+    // Parameters to pass to OAuth 2.0 endpoint.
+    const params = new Map<string, string>([
+      ['client_id', this.#clientId],
+      ['redirect_uri', this.#redirectUri],
+      ['response_type', 'code'],
+      ['scope', this.#oauthScopes.join(' ')],
+      ['include_granted_scopes', 'true'],
+      ['state', 'pass-through value'],
+    ]);
+
+    // Add form parameters as hidden input values.
+    params.forEach((value, key) => {
+      const input = document.createElement('input');
+      input.setAttribute('type', 'hidden');
+      input.setAttribute('name', key);
+      input.setAttribute('value', value);
+      form.appendChild(input);
+    });
+
+    // Add form to page and submit it to open the OAuth 2.0 endpoint.
+    document.body.appendChild(form);
+    form.submit();
+  }
+
+  async #runHttpServer(): Promise<void> {
+    const server = await this.#socketServer.opened;
+    const connections = server.readable.getReader();
+
+    while (true) {
+      const { value: connection, done } = await connections.read();
+
+      // Send the connection to the callback
+      if (connection) {
+        this.#connectionReceived(connection);
+      }
+
+      // Release the connection if we're done
+      if (done) {
+        connections.releaseLock();
+        break;
+      }
+    }
+
+    // Wait for the server to be closed
+    await this.#socketServer.closed;
+  }
+
+  async #connectionReceived(connection: TCPSocket): Promise<void> {
+    return connection.opened.then(async (socket) => {
+      const reader = socket.readable.getReader();
+      const value = await this.#readStream(reader);
+      if (value == undefined) {
+        return;
+      }
+
+      const { success, html } = this.#generateResponse(value);
+      await this.#writeStream(socket, html);
+
+      connection.close();
+
+      if (success) {
+        this.#accessCode = await this.#getAccessCode();
+        if (this.#promiseResolver != undefined) {
+          this.#promiseResolver(this.#accessCode);
+        }
+      }
+    });
```

Comment:
Please rewrite this to use the `async` function instead of returning a promise and a `.then` chain.

Reply:
Done.